botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+         * ``os/linux md/arch#x86_64``
+        """
+        if self._platform_name is None:
+            return [UserAgentComponent('os', 'other')]
+
+        plt_name_lower = self._platform_name.lower()
+        if plt_name_lower in _USERAGENT_ALLOWED_OS_NAMES:
+            os_family = plt_name_lower
+        elif plt_name_lower in _USERAGENT_PLATFORM_NAME_MAPPINGS:
+            os_family = _USERAGENT_PLATFORM_NAME_MAPPINGS[plt_name_lower]
+        else:
+            os_family = None
+
+        if os_family is not None:
+            return [
+                UserAgentComponent('os', os_family, self._platform_version)
+            ]
+        else:
+            return [
+                UserAgentComponent('os', 'other'),
+                UserAgentComponent(
+                    'md', self._platform_name, self._platform_version
+                ),
+            ]
+
+    def _build_language_metadata(self):
+        """
+        Build the language components of the User-Agent header string.
+
+        Returns the Python version in a component with prefix "lang" and name
+        "python". The Python implementation (e.g. CPython, PyPy) is returned as
+        separate metadata component with prefix "md" and name "pyimpl".
+
+        String representation of an example return value:
+        ``lang/python/3.10.4 md/pyimpl/Cpython``
+        """
+        lang_md = [
+            UserAgentComponent('lang', 'python', self._python_version),
+        ]
+        if self._python_implementation:
+            lang_md.append(
+                UserAgentComponent('md', 'pyimpl', self._python_implementation)
+            )
+        return lang_md
+
+    def _build_execution_env_metadata(self):
+        """
+        Build the execution environment component of the User-Agent header.
+
+        Returns a single component prefixed with "exec-env", usually sourced
+        from the environment variable AWS_EXECUTION_ENV.
+        """
+        if self._execution_env:
+            return [UserAgentComponent('exec-env', self._execution_env)]
+        else:
+            return []
+
+    def _build_feature_metadata(self):
+        """
+        Build the features componentes of the User-Agent header string.
+
+        Botocore currently does not report any features. This may change in a
+        future version.
+        """
+        return []
+
+    def _build_config_metadata(self):
+        """
+        Build the configuration components of the User-Agent header string.
+
+        Returns a list of components with prefix "cfg" followed by the config
+        setting name and its value. Tracked configuration settings may be
+        added or removed in future versions.
+        """
+        if not self._client_config or not self._client_config.retries:
+            return []
+        retry_mode = self._client_config.retries.get('mode')
+        cfg_md = [UserAgentComponent('cfg', 'retry-mode', retry_mode)]
+        if self._client_config.endpoint_discovery_enabled:
```

Comment:
This seems unfortunately manual. It's a one off for now, but hopefully we have a more robust way of handling in the future. We don't need to engineer it now but if we expand it may be worth wiring something in the ConfigProvider itself.

Reply:
Agreed, once there is a longer list of config settings to track, a more generic solution to this would be desirable. Our organically grown collection of interfaces won't make this easier because some things others might consider "configuration" comes from a mix of sources in Botocore including the config provider, config store, session object properties, create_client() arguments, and so on...